### PR TITLE
FCM: Fix crash on iOS 14 when multiple app delegate completion methods are called

### PR DESCRIPTION
**Pull request by mikeger**
## Background

Since iOS 7 it is possible to execute the code in the background for iOS apps that receive silent push notifications. Once such push (most likely containing the `content-available` flag) is received, if the application is in the background, it has 30 seconds to complete the background activity and tell iOS that it's done by calling the `completion` block. More info [on application(_:didReceiveRemoteNotification:fetchCompletionHandler:) here](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623013-application)

## Issue we observed

We discovered that on iOS 14 calling the completion block multiple times is leading to a crash. Based on my limited debugging capabilities, it looks like the API is now using dispatch groups to trace the moment when the completion is called, and in the completion block it invokes `dispatch_group_leave`. Once the completion block is called multiple times, the iOS is going to crash since this behavior is unexpected (calling `dispatch_group_leave` more times than `dispatch_group_enter`).

More people complained about the same issue here https://github.com/invertase/react-native-firebase/issues/3944. 

Worth noticing that in our experience the crash is only happening when the app is actually running in the foreground and in case the FCM is used together with the Braze SDK (they are also using `GULAppDelegateSwizzler`).

## Proposed solution

> Similia similibus curantur

It is sufficient to wrap the final completion block provided by iOS into yet another `dispatch_group` that would be entered for every observer of the `application(_:didReceiveRemoteNotification:fetchCompletionHandler:)` and left when the observer is calling to the provided callback.

### CODE REVIEW

**Comment by mikeger** [comment]
@googlebot I signed it!

**Comment by mikeger** [comment]
I am afraid there are some CI errors in GitHub Actions, seems like something unrelated to my change. Should I rebase it on your recent `release-xxx` branch?

**Comment by paulb777** [comment]
@mikeger Thanks for the PR! The CI failure is a result of GitHub Actions changing its default Xcode from 11.7 to 12.0.1 yesterday and being addressed in #6865.  You can disregard for now.  master is the correct target for this PR.

Assigning for review to @maksymmalyhin for the GoogleUtilities changes and @chliangGoogle for the FirebaseMessaging changes.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m
Comment (maksymmalyhin) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m:888]: It looks like calling `dispatch_group_notify` on an empty group will lead to scheduling the block straight away. I think it should be moved to after line 916.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m
Comment (maksymmalyhin) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m:912]: The current implementation doesn't add the originally implemented method to the group. I think we should do the same for it: add enter/leave group calls and wrap the completion. Potentially we may use a single completion handler wrapper to be passed to all interceptors and the original implementations.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m
Comment (maksymmalyhin) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m:898]: I think we can improve combining the fetch results. Perhaps if there is at least on `UIBackgroundFetchResultNewData` the combined result should be also `UIBackgroundFetchResultNewData`. Also I would pass `UIBackgroundFetchResultFailed` only if all calls failed.

**Review by maksymmalyhin — CHANGES_REQUESTED** [review]
Great idea, thank you for your contribution!

I added a couple comments to adjust the logic a bit. Feel free to ask clarification questions as needed. 

Also we will need to add unit tests for these changes before merging. Please let us know if you help with it.

**Comment by mikeger** [comment]
@maksymmalyhin thanks for the review, all valid comments.

**Comment by paulb777** [comment]
#6865 has merged so rebasing or merging master should now enable a green CI run.

**Comment by AlexPritchin** [comment]
Hello. Are there any updates here? I'm facing exact same issue in Firebase 6.18, 6.31.1, 6.34.0 and even 7.0.0. However I'm not using Braze SDK. According to Crashlytics 80% of such crashes occur on iOS 14.1. Others on other versions of iOS 14. So I'm waiting for this fix in next release version of Firebase. BTW the project I'm supporting now is written on Objective C.

**Comment by maksymmalyhin** [comment]
@mikeger Will you have time to address the comments? We may finish your PR if not. We will make sure your contribution will be visible in Firebase repo. Please let us know.

**Comment by mikeger** [comment]
@maksymmalyhin hey, I would address the comments today. I got stuck since the tests (`./test.sh`) were not running for me (build failed), I will retry rebased on the current master.

**Comment by maksymmalyhin** [comment]
@mikeger Ah, `./test.sh` is an obsolete script that should've been removed. Sorry about that. Currently we mostly use the following ways to run tests:
1. Follow the [development guide](https://github.com/firebase/firebase-ios-sdk#development) and run tests from the generated project. This is the most convenient way for active development.
2. Running `pod lib lint` for a podspec. See CI for examples, e.g. for [FirebaseMessaging](https://github.com/firebase/firebase-ios-sdk/blob/2d39b6c6f6b19c5bf8207d626a581e5d8ac0459e/.github/workflows/messaging.yml#L54) or [GoogleUtilities](https://github.com/firebase/firebase-ios-sdk/blob/2d39b6c6f6b19c5bf8207d626a581e5d8ac0459e/.travis.yml#L176). This is useful to debug CI issues locally.

**Comment by mikeger** [comment]
@maksymmalyhin thanks so much for the testing tip, this allowed me to progress with the tests. Please re-review the changes, I updated it based on your comments.

File: GoogleUtilities/Tests/Unit/Swizzler/GULAppDelegateSwizzlerTest.m
Comment (maksymmalyhin) [inline GoogleUtilities/Tests/Unit/Swizzler/GULAppDelegateSwizzlerTest.m:1218]: nit: Looks like these 3 tests are mostly identical except the background fetch result. Could you please extract the test logic to a helper function like `assertApplicationDidReceiveRemoteNotificationWithCompletionCompletionIsCalledOnceWithResult:(UIBackgroundFetchResult)` to reduce code duplication.

File: GoogleUtilities/Tests/Unit/Swizzler/GULAppDelegateSwizzlerTest.m
Comment (maksymmalyhin) [inline GoogleUtilities/Tests/Unit/Swizzler/GULAppDelegateSwizzlerTest.m:1139]: It would be nice to add tests for cases with different fetch results or at least add a `TODO` to add them.

**Review by maksymmalyhin — COMMENTED** [review]
Great, thank you for addressing the comments! It mostly looks good to me. I added a couple minor comments. The comments are mostly optional and I'm fine proceeding with proceeding as is.

@chliangGoogle Could you please take a look at FCM part.

**Review by maksymmalyhin — APPROVED** [review]

**Comment by mikeger** [comment]
@maksymmalyhin updated, thanks for the comments!

File: GoogleUtilities/Tests/Unit/Swizzler/GULAppDelegateSwizzlerTest.m
Comment (mikeger) [reply GoogleUtilities/Tests/Unit/Swizzler/GULAppDelegateSwizzlerTest.m]: _I thought it's paid by the number of lines_

**Review by maksymmalyhin — APPROVED** [review]
👍

**Review by maksymmalyhin — CHANGES_REQUESTED** [review]
I think we've missed on place where we need to call the completion handler after the change: could you please add it to [auth](https://github.com/firebase/firebase-ios-sdk/blob/57123ae0d5578733e4ba2a1abb7c15893651c886/FirebaseAuth/Sources/Auth/FIRAuth.m#L1501) as well.

**Comment by mikeger** [comment]
@maksymmalyhin added, to my understanding there it's sufficient to always indicate `NoData`.

**Review by chliangGoogle — APPROVED** [review]
LGTM on messaging

**Review by maksymmalyhin — APPROVED** [review]

**Comment by maksymmalyhin** [comment]
All looks good to me, I'm going to merge the PR on CI green.

**Comment by maksymmalyhin** [comment]
@mikeger Thank you again for your contribution! 

We will try to include the fix into 7.1.0 release.

**Pull request merged**
